Code/Framework/AzToolsFramework/AzToolsFramework/Prefab/Undo/PrefabUndoDelete.cpp

```diff
@@ -0,0 +1,96 @@
+
+            PrefabDomReference cachedOwningInstanceDom = focusedInstance.GetCachedInstanceDom();
+
+            // Removes entity DOMs.
+            for (const AZStd::string& entityAliasPath : entityAliasPathList)
+            {
+                if (entityAliasPath.empty())
+                {
+                    AZ_Error("Prefab", false, "PrefabUndoDeleteEntity::Capture - Entity alias path is empty. Skips removal.");
+                    continue;
+                }
+
+                PrefabUndoUtils::AppendRemovePatch(m_redoPatch, entityAliasPath);
+
+                PrefabDomPath entityDomPathInFocusedTemplate(entityAliasPath.c_str());
+                const PrefabDomValue* entityDomInFocusedTemplate = entityDomPathInFocusedTemplate.Get(focusedTempalteDom);
+                PrefabUndoUtils::AppendAddEntityPatch(m_undoPatch, *entityDomInFocusedTemplate, entityAliasPath);
+
+                // Preemptively updates the cached DOM to prevent reloading instance.
+                if (cachedOwningInstanceDom.has_value())
+                {
+                    PrefabUndoUtils::UpdateCachedOwningInstanceDomWithRemoval(cachedOwningInstanceDom, entityAliasPath);
+                }
+            }
+
+            // Updates parent entity DOMs.
+            for (const AZ::Entity* parentEntity : parentEntityList)
+            {
+                // This is actually a from-focused path because focused instance is the owning instance in the current workflow.
+                const AZStd::string parentEntityAliasPath = m_instanceToTemplateInterface->GenerateEntityAliasPath(parentEntity->GetId());
+                PrefabDomPath parentEntityAliasDomPathFromFocused(parentEntityAliasPath.c_str());
+
+                // This scope is added to limit usage and ensure DOM is not modified when it is being used.
+                {
+                    // DOM value pointers can't be relied upon if the original DOM gets modified after pointer creation.
+                    const PrefabDomValue* parentEntityDomInFocusedTemplate = parentEntityAliasDomPathFromFocused.Get(focusedTempalteDom);
+                    if (!parentEntityDomInFocusedTemplate)
+                    {
+                        AZ_Error("Prefab", false, "PrefabUndoDeleteEntity::Capture - Cannot retrieve parent entity DOM from focused template.");
+                        continue;
+                    }
+
+                    PrefabDom parentEntityDomAfterRemovingChildren;
+                    m_instanceToTemplateInterface->GenerateDomForEntity(parentEntityDomAfterRemovingChildren, *parentEntity);
+
+                    PrefabUndoUtils::AppendUpdateEntityPatch(
+                        m_redoPatch, *parentEntityDomInFocusedTemplate, parentEntityDomAfterRemovingChildren, parentEntityAliasPath);
+                    PrefabUndoUtils::AppendUpdateEntityPatch(
+                        m_undoPatch, parentEntityDomAfterRemovingChildren, *parentEntityDomInFocusedTemplate, parentEntityAliasPath);
```

Comment:
I'm wondering if we can optimize this by generating the remove patches manually rather than doing a "GenerateDomForEntity".

Reply:
AFAIK, AppendUpdateEntityPatch is doing two jobs:

1. Update component properties for entity
2. Update parent entity's child sort order

For the first job, I think it might be yes. But we need to do some profiling first to see if it brings noticeable improvement, because imo generating entity DOM is acceptable compared to generating an instance DOM. Another pain point would be how we can locate a particular component rather than doing a linear search.

For the second job, I think it might be hard. This type of patches can be >1 depending on which entity you're updating in the N children list. If it is at the first position, it would be N (or N-1) patches. It is like inserting an element to vector like what `m_numbers.insert(0, newNumber)` does to move elements behind forward by 1 position.

So due to multiple factors, we are still generating entity DOM from serialization and doing diff to find what needs to updated. But I think we already started being smart about this for updates like adding/removing an entity DOM.

Reply:
That's fair. Thanks for the details.